src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -566,11 +566,23 @@ private function validateAndDenormalizeLegacy(array $types, string $currentClass
                     return (float) $data;
                 }
 
-                if ((LegacyType::BUILTIN_TYPE_FALSE === $builtinType && false === $data) || (LegacyType::BUILTIN_TYPE_TRUE === $builtinType && true === $data)) {
-                    return $data;
-                }
-
-                if (('is_'.$builtinType)($data)) {
+                $dataMatchesExpectedType = match ($builtinType) {
+                    LegacyType::BUILTIN_TYPE_ARRAY => \is_array($data),
+                    LegacyType::BUILTIN_TYPE_BOOL => \is_bool($data),
+                    LegacyType::BUILTIN_TYPE_CALLABLE => \is_callable($data),
+                    LegacyType::BUILTIN_TYPE_FALSE => false === $data,
+                    LegacyType::BUILTIN_TYPE_FLOAT => \is_float($data),
+                    LegacyType::BUILTIN_TYPE_INT => \is_int($data),
+                    LegacyType::BUILTIN_TYPE_ITERABLE => is_iterable($data),
+                    LegacyType::BUILTIN_TYPE_NULL => null === $data,
+                    LegacyType::BUILTIN_TYPE_OBJECT => \is_object($data),
+                    LegacyType::BUILTIN_TYPE_RESOURCE => $data,
+                    LegacyType::BUILTIN_TYPE_STRING => \is_string($data),
+                    LegacyType::BUILTIN_TYPE_TRUE => true === $data,
+                    default => false,
+                };
```

Comment:
Would worth a private method maybe? As you're using the same logic twice?

Reply:
In fact, the legacy type doesn't support mixed here. So I have reverted these changes.

Reply:
@stof You are right, but we need to fix this on the `5.4` branch (see #57341).